### ISSUE TITLE
[BUG] - involute_gears are unusable

### ISSUE THREAD

**Issue by ftlpilot** [revarbat/BOSL#73]
involute_gears. are unusable - the 2D shape is WRONG and doesn't extrude to 3D - this was supposedly fixed, but it' still there today - here's a screenshot of the example code.

![bosl-bugs](https://user-images.githubusercontent.com/8108587/94505455-7941eb80-024e-11eb-8359-2b860c94cb8c.jpg)

This would be a really nice library - IF IT ACTUALLY WORKED !!!

**Comment by mjparme** [comment]
I have used involute-gears just fine before. What do you mean it is unusable?

**Comment by ftlpilot** [comment]
The inside is hollow and they can’t be rendered – look **CAREFULLY** at the pic I included, there is just an outer skin – THE CENTER IS MISSING, also this problem has been reported before.

 

Don’t rely on using them before – CHECK THEM NOW – THEY DON’T FUCKING WORK !!!!!!!!!!!!!

**Comment by revarbat** [comment]
Took a bit to figure this one out. `gear2d()` was using tooth profiles that had very slightly overlapping lines between consecutive teeth.  When that path was converted to a polygon via `polygon()`, it *looked* fine, but internally had a bad geometry, so that `linear_extrude()` failed to extrude it correctly.

Fixed in 56756cc137281b4319dea5083262f22c7aa0b804

**Comment by ftlpilot** [comment]
Thankx – much appreciated, I’ll try and work in a mention in my next DIYODE article about gears, probably Dec or Jan issue.

**Comment by ftlpilot** [comment]
Thanks again – gears work perfectly now – I’ll give your Library a mention in my next DIYODE article about Gears, in the next month or two.

 

FYI, the 1st Gears article is at  <https://diyodemag.com/education/exploring_3d_3d_printed_gears_part_1> https://diyodemag.com/education/exploring_3d_3d_printed_gears_part_1

**Comment by revarbat** [comment]
Honestly, for gears code I'd use the involute_gears.scad file in the BOSL2 library, not BOSL1.   In BOSL2 I did a complete rewrite that renders faster and more accurately.  Also it has proper support for spiral bevel gears and worm gears.

**Comment by ftlpilot** [comment]
Thankx – until you mentioned it I didn’t know there WAS a V2 – perhaps mentioning BOSL2 on https://www.openscad.org/libraries.html would be a good idea, at least when it’s out of alpha/beta testing

**Comment by ftlpilot** [comment]
As you suggested, I tried the the involute_gears.scad file in the BOSL2 library – it’s pretty much unusable, firstly there was no include/use statement for attachments, when I inserted one in my code – this was the result

 

WARNING: Ignoring unknown function 'is_vector', in file ../../../../Users/Geoff/Documents/OpenSCAD/libraries/BOSL2/attachments.scad, line 220.

ERROR: Assertion 'is_vector(cp)' failed in file ../../../../Users/Geoff/Documents/OpenSCAD/libraries/BOSL2/attachments.scad, line 220

TRACE: called by 'attach_geom', in file ../../../../Users/Geoff/Documents/OpenSCAD/libraries/BOSL2/attachments.scad, line 910.

TRACE: called by 'attachable', in file ../../../../Users/Geoff/Documents/OpenSCAD/libraries/BOSL2/Involute_gears.scad, line 700.

TRACE: called by 'gear', in file Geoffs-Dual-Geared-Extruder-All-3D-printed-Mk21v.scad, line 1590.

TRACE: called by 'translate', in file Geoffs-Dual-Geared-Extruder-All-3D-printed-Mk21v.scad, line 1589.

TRACE: called by 'union', in file Geoffs-Dual-Geared-Extruder-All-3D-printed-Mk21v.scad, line 1569.

TRACE: called by 'difference', in file Geoffs-Dual-Geared-Extruder-All-3D-printed-Mk21v.scad, line 1568.

TRACE: called by 'Geoffs_double_helix', in file Geoffs-Dual-Geared-Extruder-All-3D-printed-Mk21v.scad, line 1538.

TRACE: called by 'rotate', in file Geoffs-Dual-Geared-Extruder-All-3D-printed-Mk21v.scad, line 1537.

TRACE: called by 'translate', in file Geoffs-Dual-Geared-Extruder-All-3D-printed-Mk21v.scad, line 1536.

TRACE: called by 'color', in file Geoffs-Dual-Geared-Extruder-All-3D-printed-Mk21v.scad, line 1535.

TRACE: called by 'union', in file Geoffs-Dual-Geared-Extruder-All-3D-printed-Mk21v.scad, line 1521.

TRACE: called by 'difference', in file Geoffs-Dual-Geared-Extruder-All-3D-printed-Mk21v.scad, line 1520.
